tests/test_request_id.py

```diff
@@ -1,77 +1,453 @@
+from flask import request
+from itertools import chain, product
 import mock
-from werkzeug.test import EnvironBuilder
+import pytest
 
-from dmutils import request_id
-from dmutils.request_id import CustomRequest
+from dmutils.request_id import init_app as request_id_init_app
 
 
-def test_get_request_id_from_request_id_header():
-    builder = EnvironBuilder()
-    builder.headers['DM-REQUEST-ID'] = 'from-header'
-    builder.headers['DOWNSTREAM-REQUEST-ID'] = 'from-downstream'
-    request = CustomRequest(builder.get_environ())
+_GENERATED_TRACE_VALUE = "di5ea5e5deadbeefbaadf00dabadcafe"
 
-    request_id = request._get_request_id('DM-REQUEST-ID',
-                                         'DOWNSTREAM-REQUEST-ID')
 
-    assert request_id == 'from-header'
+_trace_id_related_params = (
+    (
+        # extra_config
+        {
+            "DM_REQUEST_ID_HEADER": "DM-REQUEST-ID",
+            "DM_DOWNSTREAM_REQUEST_ID_HEADER": "DOWNSTREAM-REQUEST-ID",
+        },
+        # extra_req_headers
+        (
+            ("DM-REQUEST-ID", "from-header",),
+            ("DOWNSTREAM-REQUEST-ID", "from-downstream",),
+        ),
+        # expected_trace_id
+        "from-header",
+        # expect_uuid_call
+        False,
+        # expected_onwards_req_headers
+        {
+            "DM-REQUEST-ID": "from-header",
+            "DOWNSTREAM-REQUEST-ID": "from-header",
+        },
+        # expected_resp_headers
+        {
+            "DM-REQUEST-ID": "from-header",
+            "DOWNSTREAM-REQUEST-ID": "from-header",
+        },
+    ),
+    (
+        # extra_config
+        {
+            "DM_REQUEST_ID_HEADER": "DM-REQUEST-ID",
+            "DM_DOWNSTREAM_REQUEST_ID_HEADER": "DOWNSTREAM-REQUEST-ID",
+        },
+        # extra_req_headers
+        (
+            ("DOWNSTREAM-REQUEST-ID", "from-downstream",),
+        ),
+        # expected_trace_id
+        "from-downstream",
+        False,
+        # expected_onwards_req_headers
+        {
+            "DM-REQUEST-ID": "from-downstream",
+            "DOWNSTREAM-REQUEST-ID": "from-downstream",
+        },
+        # expected_resp_headers
+        {
+            "DM-REQUEST-ID": "from-downstream",
+            "DOWNSTREAM-REQUEST-ID": "from-downstream",
+        },
+    ),
+    (
+        # extra_config
+        {
+            "DM_REQUEST_ID_HEADER": "DM-REQUEST-ID",
+            "DM_DOWNSTREAM_REQUEST_ID_HEADER": "",
+        },
+        # extra_req_headers
+        (),
+        # expected_trace_id
+        _GENERATED_TRACE_VALUE,
```

Comment:
Nice one on abstracting `_GENERATED_TRACE_VALUE`. Definitely good for readability